### PR TITLE
[mlir][TilingInterface] Update documentation for `TilingInterface.td`.

### PR DESCRIPTION
**Pull request by MaheshRavishankar**

### CODE REVIEW

File: mlir/include/mlir/Interfaces/TilingInterface.td [hunk 1]
```diff
@@ -18,9 +18,46 @@ include "mlir/IR/OpBase.td"
...
+    loop structure, and call into the methods of the interface to be able to
+    tile any operation that implements the interface.
+    
+    This interface is also meant to help with "tile and fuse", i.e. the process
+    of fusing a producer with a consumer by
+      a) Tiling the consumer
+      b) Based on the tile of the producer used by the tiled consumer,
+         materialize the tiled implementation of a producer to generate that
+         tile (and use it immediately in the consumer)
+    You could also fuse a consumer with with a producer by
```
Comment (qedawkins) [inline mlir/include/mlir/Interfaces/TilingInterface.td:36]:
```suggestion
    You could also fuse a consumer with a producer by
```

File: mlir/include/mlir/Interfaces/TilingInterface.td [hunk 2]
```diff
@@ -49,19 +86,18 @@ def TilingInterface : OpInterface<"TilingInterface"> {
...
           - `offsets` provides the offset of the tile in the coordinate system
             of the original iteration space, i.e., if an iteration space
-            dimension had non-zero offset, it must be included in the offset
+            dimension had non-zero offset, it will be included in the offset
             provided here (as opposed to zero-based offset "relative" to the
             iteration space).
           - `sizes` provides the size of the tile.
-
-          The method returns the operation that is the tiled
-          implementation.
```
Comment (qedawkins) [inline mlir/include/mlir/Interfaces/TilingInterface.td:64]: Add a brief description of the return type (TilingResult) here? Explain that the implementation must return the value(s) corresponding to the result tiles, not the full operation.

File: mlir/include/mlir/Interfaces/TilingInterface.td [hunk 3]
```diff
@@ -76,11 +112,32 @@ def TilingInterface : OpInterface<"TilingInterface"> {
...
-          Method to return the position of the result tile computed by the tiled operation.
+          Method to return the position of the result tile computed by the
+          tiled operation.
+
+          For operations that return a value (typically a value of type
+          `RankedTensorType`), the generated tiled computation has to also
+          recompute a replacement for the results of the original operation.
+          The tiled implementation of the operation returns a tile of the
+          result(s). This methods returns information about what part of the
+          result tensor is computed by the tiled implementation. The manner in
```
Comment (qedawkins) [inline mlir/include/mlir/Interfaces/TilingInterface.td:123]:
```suggestion
          result is computed by the tiled implementation. The manner in
```

File: mlir/include/mlir/Interfaces/TilingInterface.td [hunk 4]
```diff
@@ -76,11 +112,32 @@ def TilingInterface : OpInterface<"TilingInterface"> {
...
+          which these tiles get put together to get the final result is upto
+          the surrounding loop construct.
```
Comment (qedawkins) [inline mlir/include/mlir/Interfaces/TilingInterface.td:125]: For memref types
```suggestion
          the surrounding loop construct and the semantics of the result types.
```

File: mlir/include/mlir/Interfaces/TilingInterface.td [hunk 5]
```diff
@@ -76,11 +112,32 @@ def TilingInterface : OpInterface<"TilingInterface"> {
...
+          - `resultNumber` is the result number of the original operation
```
Comment (qedawkins) [inline mlir/include/mlir/Interfaces/TilingInterface.td:126]: `result number` seems like non-standard notation to me. Maybe `index of the OpResult`?

File: mlir/include/mlir/Interfaces/TilingInterface.td [hunk 6]
```diff
@@ -98,51 +155,66 @@ def TilingInterface : OpInterface<"TilingInterface"> {
...
         /*desc=*/[{
-          Method to return the tile of the iteration domain where
-          values from the given tile of the operand are used.
+          Method to generate the code that produces a tile of the result.
+
+          This method is required to allow operations to be "tiled and fused"
+          with an (already tiled) consumer. Typically, for two operations with
+          producer -> consumer relation ship, to compute a tile of the
+          consumer a `slice` of the producer is needed. This method allows
+          computing that slice of the produce in-place, there-by "fusing"
```
Comment (qedawkins) [inline mlir/include/mlir/Interfaces/TilingInterface.td:164]:
```suggestion
          computing that slice of the producer in-place, thereby "fusing"
```

File: mlir/include/mlir/Interfaces/TilingInterface.td [hunk 7]
```diff
@@ -98,51 +155,66 @@ def TilingInterface : OpInterface<"TilingInterface"> {
...
+          method generates the tile of the consumer that uses exactly this
+          produced tile. In some sense it is the "reverse" of
+          `generateResultTileValue`.
+          - `operandNumber` is the result of the producer used by the consumer.
+          - `offsets` is the offset of the slice of the producer result used by
+            the tiled implementation of the consumer.
+          - `sizes` is the size of the slice of the producer result used by the
+            consumer.
+          If it is illegal to fuse with a producer along the given operand for
+          an operation, the implementation should return a failure.
```
Comment (qedawkins) [inline mlir/include/mlir/Interfaces/TilingInterface.td:211]: Add a note that the implementation can assume that the given operand tile encompasses the entire iteration space (of that operand) in the aggregate of the surrounding loop construct. In other words, if there is a bijection from operand tile to iteration space, then the generated result tile will also encompass the entire untiled operation result. It is the responsibility of the caller to ensure that fusion happens only when it is valid to do so.

File: mlir/include/mlir/Interfaces/TilingInterface.td [hunk 8]
```diff
@@ -152,38 +224,69 @@ def TilingInterface : OpInterface<"TilingInterface"> {
...
+          ```
+
+          i.e, when fusing the consumer, the replacement for the result of the
+          consumer needs to be returned to replace the uses of the consumer.
+          For the tile+fuse algorithm to do this it needs information about
+          which tile of the iteration space encompasses all uses of the tile
+          produced and use that to compute what are the results produced.
+
+          Note that this method is only used as a way to implement the
+          transformation. It does not provide gaurantees on whether such a
```
Comment (qedawkins) [inline mlir/include/mlir/Interfaces/TilingInterface.td:274]:
```suggestion
          transformation. It does not provide guarantees on whether such a
```

File: mlir/include/mlir/Interfaces/TilingInterface.td (hunk 7 above)
Comment (qedawkins) [reply mlir/include/mlir/Interfaces/TilingInterface.td:211]: Edit: This comment might be better suited for `getIterationDomainTileFromOperandTile`

**Review by qedawkins — COMMENTED** [review]
Nice, thanks for the quick turnaround on this

File: mlir/include/mlir/Interfaces/TilingInterface.td (hunk 3 above)
Comment (MaheshRavishankar) [reply mlir/include/mlir/Interfaces/TilingInterface.td:123]: I think this is expected to mostly work on tensors. For memref-based ops, there are no return values.

File: mlir/include/mlir/Interfaces/TilingInterface.td (hunk 4 above)
Comment (MaheshRavishankar) [reply mlir/include/mlir/Interfaces/TilingInterface.td]: What were you trying to communicate with this addition?

File: mlir/include/mlir/Interfaces/TilingInterface.td (hunk 5 above)
Comment (MaheshRavishankar) [reply mlir/include/mlir/Interfaces/TilingInterface.td:132]: `OpResult` has a method called `resultNumber`... so pretty standard?

File: mlir/include/mlir/Interfaces/TilingInterface.td (hunk 7 above)
Comment (MaheshRavishankar) [reply mlir/include/mlir/Interfaces/TilingInterface.td:217]: Added it to `getIterationDomainTileFromOperandTile`.

File: mlir/include/mlir/Interfaces/TilingInterface.td (hunk 5 above)
Comment (qedawkins) [reply mlir/include/mlir/Interfaces/TilingInterface.td:132]: Oh I see, I had thought it was `index`. `resultNumber` is fine then.

File: mlir/include/mlir/Interfaces/TilingInterface.td (hunk 4 above)
Comment (qedawkins) [reply mlir/include/mlir/Interfaces/TilingInterface.td]: For memref tiling there is no insertion, but based on your previous comment this is never used for memrefs, so you can ignore this suggestion.

**Review by qedawkins — APPROVED** [review]

File: mlir/include/mlir/Interfaces/TilingInterface.td [hunk 9]
```diff
@@ -18,9 +18,46 @@ include "mlir/IR/OpBase.td"
...
+    You could also fuse a consumer with a producer by
+      a) Tiling the producer
+      b) Based on the tile produced, materialize the tiled implementation of
+         a consumer that uses this tile.
+    Note that the tile and fuse does not make any calculation on whether it
+    is "profitable to do this", but simply provides a mechansim to implement
+    the transformation when such a fusion is needed by the caller.
+
+    For any operation to be tilable, an operation has to implement the
+    following two methods (see description below)
```
Comment (ftynse) [inline mlir/include/mlir/Interfaces/TilingInterface.td:45]: there are four methods listed

**Pull request merged**